src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,345 @@
+    }
+
+    CBlockIndex* mine_block(bool signal)
+    {
+        CBlockHeader header;
+        header.nVersion = signal ? m_signal : m_no_signal;
+        header.nTime = m_start_time + m_blocks.size() * m_interval;
+        header.nBits = 0x1d00ffff;
+
+        auto current_block = std::make_unique<CBlockIndex>(header);
+        current_block->pprev = tip();
+        current_block->nHeight = m_blocks.size();
+        current_block->BuildSkip();
+
+        return m_blocks.emplace_back(std::move(current_block)).get();
+    }
+};
+
+void initialize()
+{
+    SelectParams(CBaseChainParams::MAIN);
+}
+} // namespace
+
+constexpr uint32_t MAX_TIME = 4102444800; // 2100-01-01
+
+FUZZ_TARGET_INIT(versionbits, initialize)
+{
+    const CChainParams& params = Params();
+
+    const int64_t interval = params.GetConsensus().nPowTargetSpacing;
+    assert(interval > 1); // need to be able to halve it
+    assert(interval < std::numeric_limits<int32_t>::max());
+
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    // these could be changed to be fuzzed inputs if desired
+    const int period = 32;
+    const int threshold = 29;
+    const size_t max_periods = 16;
+
+    static_assert(0 < threshold && threshold < period, "must be able to not signal!");
+    const size_t max_blocks = period * max_periods;
+
+    // pick the timestamp to switch based on a block
+    // note states will change *after* these blocks because mediantime lags
+    int start_block = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, period * (max_periods - 3));
+    int end_block = fuzzed_data_provider.ConsumeIntegralInRange<int>(start_block, period * (max_periods - 3));
+
+    const int64_t block_start_time = fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(params.GenesisBlock().nTime, MAX_TIME);
+
+    // too many blocks at 10min each might cause uint32_t time to overflow if
+    // block_start_time is at the end of the range above
+    assert(std::numeric_limits<uint32_t>::max() - MAX_TIME > interval * max_blocks);
+
+    // what values for version will we use to signal / not signal?
+    const int32_t ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+    const int32_t ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+
+    // select deployment parameters: bit, start time, timeout
+    const int bit = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, VERSIONBITS_NUM_BITS - 1);
+
+    bool always_active_test = false;
+    bool never_active_test = false;
+    int64_t start_time;
+    int64_t timeout;
+    if (fuzzed_data_provider.ConsumeBool()) {
+        start_time = block_start_time + start_block * interval;
+        timeout = block_start_time + end_block * interval;
+
+        assert(start_time <= timeout);
+
+        // allow for times to not exactly match a block
+        if (fuzzed_data_provider.ConsumeBool()) start_time += interval / 2;
+        if (fuzzed_data_provider.ConsumeBool()) timeout += interval / 2;
```

Comment:
@JeremyRubin on #21377 wrote:

> @MarcoFalke's comment is a slight misunderstanding I believe of how the fuzzer simulates time; I recall AJ saying (somewhere) that block times are steady interval in the fuzzing so that it doesn't matter (can't find that comment now tho). Is that correct?

The fuzzer picks the mtp parameters for the deployment by first picking a random block (`start_block` or `end_block`), figuring out the timestamp that block will have (same formula as `Blocks::mine_block` use for `nTime`), and then randomly bumps one or the other by half the interval, just to check it's not relying on exact equality.

(So switching fuzzing from mtp to height should just be a matter of using `start_block` and `end_block` directly)

The steady interval means there's no direct testing of what happens if blocks happen to come really quickly or really slowly, however if the fuzzer picks two blocks in the same period for start/timeout, that should pretty much simulate the same behaviour.